File: RotatedSortArray.java
// Time Complexity : O(logN)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Yes


// Your code here along with comments explaining your approach
//If my mid=target…I’m good to go, then I will check if my left is<mid then I can assume 
//that my left part is sorted, then I will check if my target is bw left and mid. Same applies for right



class RotatedSortArray {
    public int search(int[] nums, int target) {
        int l = 0,m;
        int r = nums.length-1;
        
        if(nums == null || nums.length==0){
            return -1;
        }
        
        while(l<=r){
            m = l+(r-l)/2;
            
            if(nums[m]==target){
                return m;
            }
            else if(nums[l]<=nums[m]){
                if(target>=nums[l] && target<nums[m]){
                    r=m-1;
                }
                else{
                    l=m+1;
                }
            }
            else{
                if(target>nums[m] && target<=nums[r]){
                    l=m+1;
                }
                else{
                    r=m-1; 
                }
            }
        }
        return -1;
    }
}

File: rotatedSortArray.py
# // Time Complexity : O(logn)
# // Space Complexity : O(1)
# // Did this code successfully run on Leetcode :yes

class Solution:
    def search(self, nums: list[int], target: int) -> int:
        if not nums or len(nums) == 0:
            return -1
        low = 0
        high = len(nums)-1
        
        while (low <= high):
            mid = low+(high-low)//2
            
            #I got the target
            if target == nums[mid]:
                return mid
            # if we don't get the target at mid: here when we closely look at the array we see one part of the array is always sorted compared other; either left is sorted compare to right or either way

            #checking if left is sorted
            if nums[low] <= nums[mid]:
                #checking if the element lies in left sorted array
                if target >= nums[low] and target < nums[mid]:
                    # if yes, go left
                    high = mid-1
                else:
                    #or go right if not
                    low = mid+1
            else:
                if target > nums[mid] and target <= nums[high]:
                    low = mid+1
                else:
                    high = mid-1
                    
        return -1


        